Ensure we use the ssh tunnel on action execution

dpsutton opened the pull request:
Fixes https://github.com/metabase/metabase/issues/29223

<img width="740" alt="image" src="https://user-images.githubusercontent.com/6377293/225153747-e8426d25-e164-4bef-9051-0beb8a889f9c.png">

Previous to this, we did not use ssh tunnel information and the connection would fail.

I'm gonna see if we have an existing way to run tests against ssh dbs. I see there are some tests that use a faking ssh server. Not sure if those are useful or not. But otherwise ssh stuff needs out of process help to accomplish so not sure if it's easily feasible right now

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/29228)
<!-- Reviewable:end -->

dpsutton committed:
Ensure we use the ssh tunnel on action execution

I'm gonna see if we have an existing way to run tests against ssh dbs. I
see there are some tests that use a faking ssh server. Not sure if those
are useful or not. But otherwise ssh stuff needs out of process help to
accomplish so not sure if it's easily feasible right now

dpsutton committed:
introduce new `with-unpooled-connection-spec` macro

it is identical to the `with-connection-spec-for-testing-connection`
macro but has a better name. In the future they could diverge as testing
requirements and action requirements change. But we want that
_implementation_ (which correctly uses tunnels), but we do not want a
"testing connection" slug in the usage.

In the future (or now if we prefer a different name) we can rename it,
find the usages that are not testing vs the ones that are, etc.

dpsutton commented on src/metabase/driver/sql_jdbc/connection.clj:
```diff
@@ -296,6 +296,19 @@
   [[jdbc-spec-binding [driver details]] & body]
   `(do-with-connection-spec-for-testing-connection ~driver ~details (^:once fn* [~jdbc-spec-binding] ~@body)))
 
+(defmacro with-unpooled-connection-spec
```
This is identical to `with-connection-spec-for-testing-connection`. That macro does all of the helpful stuff of ensuring we have an open tunnel and the tunnel closes after we are done. It's perfect. But we don't want that name in the action execution code. And if they ever need to change in the future, we don't have to guess which ones are testing a connection and which are using a non-pooled connection on purpose.

dpsutton committed:
tests for actions over ssh

dpsutton committed:
unused require of mb.u.ssh

had brought it in but then stuffed all of this inside of
`with-unpooled-connection-spec` so no need to worry about ssh, closing
it, ensuring the port is there, etc.

dpsutton committed:
Don't test h2 with ssh tunnel

details lack a host and it npe's in the ssh tunnel stuff. No big loss as
ssh on h2 is not supported (due to this error, and probably many more)

qnkhuat commented on test/metabase/actions_test.clj:
```diff
@@ -477,3 +485,96 @@
+      (with-open [ssh-server (basic-auth-ssh-server username password)]
+        (doseq [[correct-password? ssh-password] [[true password] [false "wrong-password"]]]
+          (with-actions-test-data-and-actions-permissively-enabled
+            (let [ssh-port (.getPort ^SshServer ssh-server)]
+              (let [details (db/select-one-field :details 'Database :id (mt/id))]
```
we can combine the 2 lets

qnkhuat reviewed (commented):
I dont know much about jdbc connections, but why cant' we use `sql-jdbc.actions/with-jdbc-transaction`?

```clojure
(defmethod driver/execute-write-query! :sql-jdbc
  [driver {{sql :query, :keys [params]} :native}]
  {:pre [(string? sql)]}
  (try
    (with-jdbc-transaction [conn (:id (qp.store/database))]
      (with-open [stmt (statement-or-prepared-statement driver conn sql params nil)]
        {:rows-affected (if (instance? PreparedStatement stmt)
                          (.executeUpdate ^PreparedStatement stmt)
                          (.executeUpdate stmt sql))}))
    (catch Throwable e
      (throw (ex-info (tru "Error executing write query: {0}" (ex-message e))
                      {:sql sql, :params params, :type qp.error-type/invalid-query}
                      e)))))
```

Modify that function to this works for me locally and it passes your tests.
This is how we get the connection to execute implication actions in `actions/perform-action!*`

calherries commented on test/metabase/actions_test.clj:
```diff
@@ -477,3 +485,96 @@
+  ;; the db on the correct port. So we do the tests twice, once expecting the correct answer, and then again with the
+  ;; wrong password for the ssh tunnel and we expect failures. This ensures that the correct result is actually going
+  ;; through the ssh tunnel
+  (mt/test-drivers (disj (mt/normal-drivers-with-feature :actions) :h2)
```
Do we support ssh tunnelling with H2? In theory someone could run an H2 database in server mode and connect to it with an SSH tunnel. See `http://h2database.com/html/features.html#auto_mixed_mode`. 

I see code supporting it [here](https://github.com/metabase/metabase/blob/26656c26ad10ed3581eb40cd682ab37cd7e612c6/src/metabase/driver/h2.clj#L496), but the "Use an SSH-tunnel" option is missing from the admin connection settings for H2 databases.

dpsutton replied:
I don't think we do. It throws an error in the tests because it's checking the host of the db details. As far as I know, h2 doesn't (ever?) have a host and it fails. To extend ssh to h2 would be a lot of work that doesn't need to happen now for this, and I suspect there isn't much demand for it anyways.

tsmacdonald commented on test/metabase/actions_test.clj:
```diff
@@ -2,14 +2,22 @@
    [metabase.util.schema :as su]
-   [schema.core :as s]))
+   [schema.core :as s]
+   [toucan.db :as db])
```
I've been trying to write new code with `t2` instead of `db`

dpsutton replied:
good call. Switching

tsmacdonald commented on test/metabase/actions_test.clj:
```diff
@@ -477,3 +485,96 @@
+                          (is (instance? Exception response) "Did not get an error with wrong password")
+                          (is (some (partial instance? org.apache.sshd.common.SshException)
+                                    (u/full-exception-chain response))
+                              "None of the errors are from ssh"))))))))))))))
```
very clever test setup

tsmacdonald reviewed (approved):
Second Cal's question about H2, but otherwise looks good to me :boat:

dpsutton committed:
Switch to toucan2 for db access in tests

dpsutton commented:
> I dont know much about jdbc connections, but why cant' we use `sql-jdbc.actions/with-jdbc-transaction`?
> 
> ```clojure
> (defmethod driver/execute-write-query! :sql-jdbc
>   [driver {{sql :query, :keys [params]} :native}]
>   {:pre [(string? sql)]}
>   (try
>     (with-jdbc-transaction [conn (:id (qp.store/database))]
>       (with-open [stmt (statement-or-prepared-statement driver conn sql params nil)]
>         {:rows-affected (if (instance? PreparedStatement stmt)
>                           (.executeUpdate ^PreparedStatement stmt)
>                           (.executeUpdate stmt sql))}))
>     (catch Throwable e
>       (throw (ex-info (tru "Error executing write query: {0}" (ex-message e))
>                       {:sql sql, :params params, :type qp.error-type/invalid-query}
>                       e)))))
> ```
> 
> Modify that function to this works for me locally and it passes your tests. This is how we get the connection to execute implication actions in `actions/perform-action!*`

I ran into this in testing. I want to find out why they got connections in two different ways. If we can hit the connection pool yeah this is way better.

dpsutton committed:
Use connection pool for custom actions connection

Was originally using `jdbc/get-connection` from the details of the
db. This was a one-off non-pooled connection that also did not reuse ssh
tunnels. I thought the one-off connection was important since it is not
read only, but that's not the case. So we can reuse the exact same
mechanism that the implicit actions are using with
`metabase.driver.sql-jdbc.actions/with-jdbc-transaction`.

dpsutton committed:
cyclic dependency

almost moved it into execute, but that gives a very inviting
`sql-jdbc.execute/with-jdbc-transaction`. Welcoming even. But
`sql-jdbc.actions/with-jdbc-transaction` lets you know its only for
transactions.
